File: LinkedList/i2-standard/rotateRight/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

// RotateRight rotates the linked list to the right by k positions.
//
// Approach:
// 1. Find the length of the list to handle cases where k > length
// 2. Calculate effective rotations: k % length (to avoid unnecessary full rotations)
// 3. Use two pointers with a gap of k nodes to find the new tail
// 4. Break the list at the new tail and reconnect to form the rotated list
//
// Time Complexity: O(n) where n is the number of nodes in the list
//   - First pass to calculate length: O(n)
//   - Second pass to find the breaking point: O(n)
//
// Space Complexity: O(1) - only using pointers, no extra data structures
func RotateRight(head *ListNode, k int) *ListNode {
	// Handle empty list edge case
	if head == nil {
		return head
	}

	// Calculate the length of the linked list
	length := 0
	for current := head; current != nil; current = current.Next {
		length++
	}

	// Calculate effective rotations (handle k > length)
	effectiveRotations := k % length

	// If no rotation needed, return original list
	if effectiveRotations == 0 {
		return head
	}

	// Find the new tail (node before the new head)
	// Use two pointers: fast pointer moves k steps ahead first,
	// then both move together until fast reaches the end
	slow := head
	fast := head

	// Move fast pointer k steps ahead
	for range effectiveRotations {
		fast = fast.Next
	}

	// Move both pointers until fast reaches the last node
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}

	// Perform the rotation:
	// 1. The new head is the node after slow
	newHead := slow.Next
	// 2. Connect the old tail (fast) to the old head
	fast.Next = head
	// 3. Break the connection at the new tail
	slow.Next = nil

	return newHead
}

File: LinkedList/i2-standard/rotateRight/main.cpp
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode* rotateRight(ListNode* head, int k) {
    if (head == nullptr || head->next == nullptr || k == 0) return head;

    // find the length of the list, and the tail
    int len = 1;
    ListNode* tail = head;
    while (tail->next != nullptr) {
      tail = tail->next;
      len++;
    }

    // rotateCount is the number of nodes to rotate
    int rotateCount = k % len;

    // if rotateCount == 0, no need to rotate
    if (rotateCount == 0) return head;

    // find the new tail
    ListNode* newTail = head;
    int findNewTailIndex = len - rotateCount - 1;
    for (int i = 0; i < findNewTailIndex; i++) {
      newTail = newTail->next;
    }

    // find the new head
    ListNode* newHead = newTail->next;

    // wire the tail to the head
    tail->next = head;

    // cut off the new tail
    newTail->next = nullptr;

    return newHead;
  }
};

// This is the solution I come up with, but it is not as good as the above one
// because it has to handle the edge case where the new start node is the tail
// node.
// And the code is not clean
class Solution {
 public:
  ListNode* rotateRight(ListNode* head, int k) {
    if (head == nullptr || head->next == nullptr || k == 0) return head;

    // find the length of the list
    int len = 0;
    ListNode* ptr = head;
    while (ptr != nullptr) {
      ptr = ptr->next;
      len++;
    }

    // find the new start node
    ptr = new ListNode(0, head);
    for (int i = 0; i < k % len; i++) {
      ptr = ptr->next;
    }
    ListNode* newStartNode = head;
    // prt->next is edge case
    while (ptr->next != nullptr && ptr->next != newStartNode) {
      newStartNode = newStartNode->next;
      ptr = ptr->next;
    }

    // wire the tail to the head
    ptr = newStartNode;
    for (int i = 0; i < len - 1; i++) {
      if (ptr->next == nullptr) {
        ptr->next = head;
      }

      ptr = ptr->next;
    }

    // cut off the new tail
    ptr->next = nullptr;

    return newStartNode;
  }
};
